python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,151 @@
+    """ID of the table’s current schema."""
+
+    partition_specs: list = Field(alias="partition-specs")
+    """A list of partition specs, stored as full partition spec objects."""
+
+    default_spec_id: int = Field(alias="default-spec-id")
+    """ID of the “current” spec that writers should use by default."""
+
+    last_partition_id: int = Field(alias="last-partition-id")
+    """An integer; the highest assigned partition field ID across all
+    partition specs for the table. This is used to ensure partition fields
+    are always assigned an unused ID when evolving specs."""
+
+    properties: dict
+    """	A string to string map of table properties. This is used to
+    control settings that affect reading and writing and is not intended
+    to be used for arbitrary metadata. For example, commit.retry.num-retries
+    is used to control the number of commit retries."""
+
+    current_snapshot_id: int = Field(alias="current-snapshot-id")
+    """ID of the current table snapshot."""
+
+    snapshots: list
+    """A list of valid snapshots. Valid snapshots are snapshots for which
+    all data files exist in the file system. A data file must not be
+    deleted from the file system until the last snapshot in which it was
+    listed is garbage collected."""
+
+    snapshot_log: list = Field(alias="snapshot-log")
+    """A list (optional) of timestamp and snapshot ID pairs that encodes
+    changes to the current snapshot for the table. Each time the
+    current-snapshot-id is changed, a new entry should be added with the
+    last-updated-ms and the new current-snapshot-id. When snapshots are
+    expired from the list of valid snapshots, all entries before a snapshot
+    that has expired should be removed."""
+
+    metadata_log: list = Field(alias="metadata-log")
+    """A list (optional) of timestamp and metadata file location pairs that
+    encodes changes to the previous metadata files for the table. Each time
+    a new metadata file is created, a new entry of the previous metadata
+    file location should be added to the list. Tables can be configured to
+    remove oldest metadata log entries and keep a fixed-size log of the most
+    recent entries after a commit."""
+
+    sort_orders: list = Field(alias="sort-orders")
+    """A list of sort orders, stored as full sort order objects."""
+
+    default_sort_order_id: int = Field(alias="default-sort-order-id")
+    """Default sort order id of the table. Note that this could be used by
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+
+class TableMetadataV1(TableMetadataCommonFields, IcebergBaseModel):
+    format_version: Literal[1] = Field(alias="format-version")
+    """An integer version number for the format. Currently, this can be 1 or 2
+    based on the spec. Implementations must throw an exception if a table’s
+    version is higher than the supported version."""
+
+    schema_: Schema = Field(alias="schema")
```

Comment:
Does this correctly translate?

When reading, we want to accept a `schema`, but convert it into a `schemas` list and `current-schema-id`. In v1, we want to write both `schema` and `schemas` / `current-schema-id` because adding that metadata is compatible.

Reply:
schema is a reserved keyword in Pydantic:

print(Model.schema())
```python
>>> table_metadata.schema()
{
	'title': 'TableMetadataV2',
	'description': 'Metadata for an Iceberg table as specified in the Apache Iceberg\nspec (https://iceberg.apache.org/spec/#iceberg-table-spec)',
	'type': 'object',
	'properties': {
		'table-uuid': {
			'title': 'Table-Uuid',
			'type': 'string',
			'format': 'uuid'
		},
		'location': {
			'title': 'Location',
			'type': 'string'
		},
		'last-updated-ms': {
			'title': 'Last-Updated-Ms',
			'type': 'integer'
		},
		'last-column-id': {
			'title': 'Last-Column-Id',
			'type': 'integer'
		},
		'schemas': {
			'title': 'Schemas',
			'type': 'array',
			'items': {
				'$ref': '#/definitions/Schema'
			}
		},
		'current-schema-id': {
			'title': 'Current-Schema-Id',
			'type': 'integer'
		},
		'partition-specs': {
			'title': 'Partition-Specs',
			'type': 'array',
			'items': {}
		},
		'default-spec-id': {
			'title': 'Default-Spec-Id',
			'type': 'integer'
		},
		'last-partition-id': {
			'title': 'Last-Partition-Id',
			'type': 'integer'
		},
		'properties': {
			'title': 'Properties',
			'type': 'object'
		},
		'current-snapshot-id': {
			'title': 'Current-Snapshot-Id',
			'type': 'integer'
		},
		'snapshots': {
			'title': 'Snapshots',
			'type': 'array',
			'items': {}
		},
		'snapshot-log': {
			'title': 'Snapshot-Log',
			'type': 'array',
			'items': {}
		},
		'metadata-log': {
			'title': 'Metadata-Log',
			'type': 'array',
			'items': {}
		},
		'sort-orders': {
			'title': 'Sort-Orders',
			'type': 'array',
			'items': {}
		},
		'default-sort-order-id': {
			'title': 'Default-Sort-Order-Id',
			'type': 'integer'
		},
		'format-version': {
			'title': 'Format-Version',
			'enum': [2],
			'type': 'integer'
		},
		'last-sequence-number': {
			'title': 'Last-Sequence-Number',
			'type': 'integer'
		}
	},
	'required': ['table-uuid', 'location', 'last-updated-ms', 'last-column-id', 'schemas', 'current-schema-id', 'partition-specs', 'default-spec-id', 'last-partition-id', 'properties', 'current-snapshot-id', 'snapshots', 'snapshot-log', 'metadata-log', 'sort-orders', 'default-sort-order-id', 'format-version', 'last-sequence-number'],
	'definitions': {
		'IcebergType': {
			'title': 'IcebergType',
			'description': "Base type for all Iceberg Types\n\nExample:\n    >>> str(IcebergType())\n    'IcebergType()'\n    >>> repr(IcebergType())\n    'IcebergType()'",
			'type': 'object',
			'properties': {}
		},
		'NestedField': {
			'title': 'NestedField',
			'description': 'Represents a field of a struct, a map key, a map value, or a list element.\n\nThis is where field IDs, names, docs, and nullability are tracked.\nExample:\n    >>> str(NestedField(\n    ...     field_id=1,\n    ...     name=\'foo\',\n    ...     field_type=FixedType(22),\n    ...     required=False,\n    ... ))\n    \'1: foo: optional fixed[22]\'\n    >>> str(NestedField(\n    ...     field_id=2,\n    ...     name=\'bar\',\n    ...     field_type=LongType(),\n    ...     is_optional=False,\n    ...     doc="Just a long"\n    ... ))\n    \'2: bar: required long (Just a long)\'',
			'type': 'object',
			'properties': {
				'id': {
					'title': 'Id',
					'type': 'integer'
				},
				'name': {
					'title': 'Name',
					'type': 'string'
				},
				'type': {
					'$ref': '#/definitions/IcebergType'
				},
				'required': {
					'title': 'Required',
					'default': True,
					'type': 'boolean'
				},
				'doc': {
					'title': 'Doc',
					'type': 'string'
				}
			},
			'required': ['id', 'name', 'type']
		},
		'Schema': {
			'title': 'Schema',
			'description': 'A table Schema\n\nExample:\n    >>> from iceberg import schema\n    >>> from iceberg import types',
			'type': 'object',
			'properties': {
				'fields': {
					'title': 'Fields',
					'type': 'array',
					'items': {
						'$ref': '#/definitions/NestedField'
					}
				},
				'schema-id': {
					'title': 'Schema-Id',
					'type': 'integer'
				},
				'identifier-field-ids': {
					'title': 'Identifier-Field-Ids',
					'type': 'array',
					'items': {
						'type': 'integer'
					}
				}
			},
			'required': ['schema-id']
		}
	}
}
"""

Therefore we need to set the alias to properly map it.

I've added code to properly map it. In v1 the schema's are optional, if it is set, I'll check if the schema is in there, otherwise, I'll add it.